### ISSUE TITLE
Expose number of running job-workers

### ISSUE THREAD

**Issue by LarsBuur** [vsivsi/meteor-job-collection#87]
I would suggest a method to help determine how many worker processes that are running at any given time. 

Currently we have one job that just updates a timestamp every ten minutes and then an external service that checks the timestamp. This way we know that at least one worker-process has run within the last ten minutes.

**Comment by vsivsi** [comment]
Hi, thanks for the suggestion. Just to clarify what you are asking for here:  do you want to know the number of running jobs?  Or the number of available worker processes?   

Getting the number of currently "running" (claimed by a worker) jobs is easy:

```js
runningJobs = jc.find({status:"running"}).count();
```

Getting a count of the number of running or available workers is a more complicated question and is mostly outside the scope of job-collection. Let me explain. Job-collection isn't involved in starting, stopping or otherwise managing remote worker processes. If you want to spin up a cluster of 50 machines to work on jobs, you need to manage that resource separately. There is no way job-collection could do this because there are so many possible interfaces to different types of compute resources. A lot of cluster managers have been written to take on this task. And so job-collection avoids any involvement with this. As the name suggests, it is focused exclusively on the jobs, not the workers. 

For these reasons, workers don't register with a job-collection, so there is no "count" of them to report. If they use authentication to connect, then Meteor's accounts system may know how many active connections there are for a given userId or userIds depending on how you manage that. But again how you choose to structure and manage your accounts, authentication and permissions is up to your application and varies widely. 

Another issue is what to count as a "worker"?  Some workers will have 32 cores working on many jobs concurrently (over an single DDP connection) while others may only work on a single job at a time. Job-collection knows nothing about any of this. Workers can start up and go down at any time (even in the middle of processing jobs), and job-collection can only detect this indirectly.  See this issue for one such example: https://github.com/vsivsi/meteor-job-collection/issues/86

If you use the `jc.processJobs()` method to create a local job queue, then it has methods to query the number of running vs idle local worker slots.  See: 

So basically, if you want to know about running jobs, it's easy. If you want to know about the pool of machines,for processors, or whatever constitutes your pool of workers, then you need to ask whatever software you use to manage those machines, because job-collection has know way of knowing what resources are on the other side of the wire.

**Comment by vsivsi** [comment]
After writing all of that, I just re-read your OP and realized you are asking for something slightly different... Are any workers available for a current job type?

The way job-collection works internally is that workers "pull" jobs from the collection by calling the `getWork()` method. If you use `processJobs()` this happens inside of that mechanism.

It would be straightforward to add a server-side property on the collection to record the last time any worker called the `getWork` method. It could even record that for each type of job requested. Would this satisfy your project's need?

**Comment by LarsBuur** [comment]
Hi Vaughn,

That would definitely satisfy my needs. 

My problem is that we have many different jobs that pulls in data from many different sources. Sometimes one process is stuck and not released. Sometimes a worker-process dies or the instance running the worker-process die. Sometimes I forget to restart it :-) after having stopped it for some reason. 

I have seen the proposed change (v. 1.2) with a timeout and that is something that we currently handle on each individual job and something that I think is a great improvement.

**Comment by vsivsi** [comment]
Okay, I'll think about how to best add this in 1.2.0  Rather than hack it in, I'm going to design a more general "status" mechanism that is extensible and can emit events when status changes so that servers can act on (or publish) those changes without needing to poll the collection.

**Comment by vsivsi** [comment]
Although in your case, I guess its the absence of change in status that would be of concern.  Hmmm.

**Comment by rhyslbw** [comment]
+1. Now really looking forward to 1.2.0 :-)

**Comment by vsivsi** [comment]
@rhyslbw  I'm open to feature requests / design ideas for this... I've opened a new issue for the discussion: https://github.com/vsivsi/meteor-job-collection/issues/89

**Comment by vsivsi** [comment]
Version 1.2.0 just released with event emitters that enable the types of stats gathering requested in the OP.  For a quick working example, look at code for the [new playground app](http://jcplayground.meteor.com/) here: https://github.com/vsivsi/meteor-job-collection-playground/blob/master/play.coffee#L384-L390

Documentation for the new events is here: https://github.com/vsivsi/meteor-job-collection#jcevents---server

**Comment by vsivsi** [comment]
Fixed long ago...
